src/ui/public/accessibility/kbn_accessible_click.js

```diff
@@ -0,0 +1,86 @@
+ *
+ * If you need to use a <div>, <p>, or <a> without the href attribute, then you need to allow
+ * them to receive focus and to respond to keyboard input. The workaround is to:
+ *
+ *   - Give the element tabindex="0" so that it can receive keyboard focus.
+ *   - Add a JavaScript onkeyup event handler that triggers element functionality if the Enter key
+ *     is pressed while the element is focused. This is necessary because some browsers do not trigger
+ *    onclick events for such elements when activated via the keyboard.
+ *   - If the item is meant to function as a button, the onkeyup event handler should also detect the
+ *     Spacebar in addition to the Enter key, and the element should be given role="button".
+ *
+ * Apply this directive to any of these elements to automatically do the above.
+ */
+
+import { accessibleClickKeys } from './accessible_click_keys';
+import { uiModules } from 'ui/modules';
+
+uiModules.get('kibana')
+.directive('kbnAccessibleClick', function ($parse) {
+  return {
+    restrict: 'A',
+    scope: {
+      kbnAccessibleClick: '&',
+    },
+    controller: $element => {
+      $element.on('keydown', e => {
+        // We don't care if the user is actually interacting with a different element.
+        if (e.currentTarget !== e.target) {
+          return;
+        }
+
+        // Prevent a scroll from occurring if the user has hit space.
+        if (accessibleClickKeys[e.keyCode]) {
+          e.preventDefault();
+        }
+      });
+    },
+    link: (scope, element, attrs) => {
+      // The whole point of this directive is to hack in functionality that native buttons provide
+      // by default.
+      const elementType = element.prop('tagName');
+
+      if (elementType === 'BUTTON') {
+        throw new Error(`kbnAccessibleClick doesn't need to be used a button.`);
+      }
+
+      if (elementType === 'A' && attrs.href !== undefined) {
+        throw new Error(`kbnAccessibleClick doesn't need to be used on a link if it has a href attribute.`);
+      }
+
+      // If the developer hasn't already specified attributes required for accessibility, add them.
+      if (attrs.tabindex === undefined) {
+        element.attr('tabindex', '0');
+      }
+
+      if (attrs.role === undefined) {
+        element.attr('role', 'button');
+      }
+
+      element.on('keyup', e => {
+        // Support keyboard accessibility by toggling display on ENTER or SPACE keypress.
+        if (accessibleClickKeys[e.keyCode]) {
+          // If a unique click handler has been specified, then call it.
+          if (attrs.kbnAccessibleClick) {
+            return scope.kbnAccessibleClick();
+          }
+
+          // Otherwise, default to the specified ng-click.
+          const ngClick = $parse(attrs.ngClick);
+
+          if (!ngClick) {
+            throw new Error('kbnAccessibleClick requires a callback or ng-click defined on its element.');
+          }
+
+          ngClick(scope.$parent);
```

Comment:
I was afraid of being too prescriptive by requiring an ng-click, but I guess you're right. That's sort of the whole idea, so I'll remove the ability to provide a handler.

Great suggestion on delegating to the ng-click via a click event, I'll implement your suggestion.

I did a little research but I couldn't find any information on which browsers may trigger click events on non-interactive elements. Hopefully we can recruit the help of an expert to help us with this (@jimgoodwin is already looking into this).